js/src/common/utils/ItemList.ts

```diff
@@ -81,51 +166,147 @@ export default class ItemList {
    * Remove an item from the list.
    */
   remove(key: string): this {
-    delete this.items[key];
+    delete this._items[key];
 
     return this;
   }
 
   /**
    * Merge another list's items into this one.
+   *
+   * The list passed to this function will overwrite items which already exist
+   * with the same key.
    */
-  merge(items: this): this {
-    for (const i in items.items) {
-      if (items.items.hasOwnProperty(i) && items.items[i] instanceof Item) {
-        this.items[i] = items.items[i];
+  merge<K>(otherList: ItemList<K>): ItemList<T | K> {
+    Object.keys(otherList._items).forEach((key) => {
+      const val = otherList._items[key];
+
+      if (val instanceof Item) {
+        (this as ItemList<T | K>)._items[key] = otherList._items[key];
       }
-    }
+    });
 
     return this;
   }
 
   /**
-   * Convert the list into an array of item content arranged by priority. Each
-   * item's content will be assigned an `itemName` property equal to the item's
-   * unique key.
+   * Convert the list into an array of item content arranged by priority.
+   *
+   * This **does not** preserve the original types of primitives and proxies
+   * all content values to make `itemName` accessible on them.
+   *
+   * **NOTE:** Modifying any objects in the final array may also update the
+   * content of the original ItemList.
+   *
+   * @see https://github.com/flarum/core/issues/3030
    */
-  toArray(): any[] {
-    const items: Item[] = [];
+  toArray(keepPrimitives?: false): (T & { itemName: string })[];
+  /**
+   * Convert the list into an array of item content arranged by priority.
+   *
+   * Content values that are already objects will be proxied and have
+   * `itemName` accessible on them. Primitive values will not have the
+   * `itemName` property accessible.
+   *
+   * **NOTE:** Modifying any objects in the final array may also update the
+   * content of the original ItemList.
+   */
+  toArray(keepPrimitives: true): (T extends object ? T & Readonly<{ itemName: string }> : T)[];
 
-    for (const i in this.items) {
-      if (this.items.hasOwnProperty(i) && this.items[i] instanceof Item) {
-        this.items[i].content = Object(this.items[i].content);
+  /**
```

Comment:
Can we get rid of this 3rd comment?

Reply:
Yes! I thought it was needed to provide JSDoc when no parameters were given, but it isn't. :)